src/main/java/org/truffleruby/extra/ffi/Pointer.java

```diff
@@ -175,11 +177,17 @@ public synchronized void enableAutorelease(FinalizationService finalizationServi
             return;
         }
 
-        finalizationService.addFinalizer(this, Pointer.class, () -> free());
+        // We must be careful here that the finalizer does not capture the Pointer itself that we'd
+        // like to finalize.
+        addFinalizer(finalizationService, this, address);
 
         autorelease = true;
     }
 
+    private static void addFinalizer(FinalizationService finalizationService, Pointer pointer, long address) {
+        finalizationService.addFinalizer(pointer, Pointer.class, () -> UNSAFE.freeMemory(address));
+    }
```

Comment:
This method still has access to the `pointer` instance.
I was thinking a method returning `() -> UNSAFE.freeMemory(address)` as a Runnable.

Reply:
Ah no, address is a field so captures the pointer object.

Reply:
I meant passing it as a local variable:
```ruby
private static Runnable finalizer(long address) {
        return () -> UNSAFE.freeMemory(address);
    }
```